File: day006_matrix_basics/related_lower_triangle_sum/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

void solve_lower_triangle_sum() {
    // Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    if (!(cin >> N)) return;

    // 1. Read and Store Matrix (O(N^2) Space)
    vector<vector<int>> matrix(N, vector<int>(N));
    long long lower_triangle_sum = 0;

    // 2. Combined Read and Sum (O(N^2) Traversal)
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            int element;
            if (!(cin >> element)) return;
            matrix[i][j] = element; // Store the matrix

            // Check the lower triangle condition: row index >= column index
            if (i >= j) {
                lower_triangle_sum += element;
            }
        }
    }

    cout << lower_triangle_sum << endl;
}

int main() {
    solve_lower_triangle_sum();
    return 0;
}

File: day006_matrix_basics/related_lower_triangle_sum/solution.py
import sys

def solve_lower_triangle_sum():
    # Read N
    try:
        N = int(sys.stdin.readline())
    except:
        return

    matrix = []
    lower_triangle_sum = 0

    # 1. Read, Store, and Sum in a single O(N^2) pass
    for i in range(N):
        try:
            row = list(map(int, sys.stdin.readline().split()))
            matrix.append(row)
        except:
            return
        
        for j in range(N):
            element = matrix[i][j]
            
            # Check the lower triangle condition: row index >= column index
            if i >= j:
                lower_triangle_sum += element

    print(lower_triangle_sum)

if __name__ == "__main__":
    solve_lower_triangle_sum()